### PR TITLE
Task/calm 25916 ansible vm crud

### PR DESCRIPTION
**Pull request by saratkumar-yb**
* Support for both name and uuid
* Allow users to increase size of images. Allow addition of empty CDROM
*  Restart VM only when needed during update
* Change in Disk & Nic structure
* Issues fixed:
   - #4 
   - #7 
   - #8 
   - #9 
   - #18 
   - #21 
   - #22

### CODE REVIEW

File: plugins/module_utils/nutanix_api_client.py [hunk 1]
```diff
@@ -137,10 +138,44 @@ def delete_vm(vm_uuid, client):
...
+            "power_state_mechanism": {
+                "mechanism": mechanism
+            }
+        }
+
+    return update_vm(vm_uuid, data, client)
+
+
+def get_image_uuid(image_name, client):
+    length = 250
```
Comment (AmitVasantSalunke) [inline plugins/module_utils/nutanix_api_client.py]: Can we use global vars to set default value 
e.g. length = DEFAULT_LENGTH

Define global vars in some other file and import it  or in same file is also fine.

File: plugins/module_utils/nutanix_api_client.py [hunk 2]
```diff
@@ -172,3 +207,133 @@ def delete_image(image_uuid, client):
     response = client.request(
         api_endpoint="v3/images/{0}".format(image_uuid), method="DELETE", data=None)
     return response.json()["status"]["execution_context"]["task_uuid"]
+
+
+def get_cluster_uuid(cluster_name, client):
+    length = 250
```
Comment (AmitVasantSalunke) [inline plugins/module_utils/nutanix_api_client.py]: Can we use global vars to set default value
e.g. length = DEFAULT_LENGTH

Define global vars in some other file and import it or in same file is also fine.

File: plugins/module_utils/nutanix_api_client.py [hunk 3]
```diff
@@ -172,3 +207,133 @@ def delete_image(image_uuid, client):
...
+                cluster_uuid.append(cluster["metadata"]["uuid"])
+
+        total_matches = cluster_list["metadata"]["total_matches"]
+        offset += length
+
+    return cluster_uuid
+
+
+def get_subnet_uuid(subnet_name, client):
+    length = 250
```
Comment (AmitVasantSalunke) [inline plugins/module_utils/nutanix_api_client.py]: Can we use global vars to set default value
e.g. length = DEFAULT_LENGTH

Define global vars in some other file and import it or in same file is also fine.

File: plugins/module_utils/nutanix_api_client.py [hunk 4]
```diff
@@ -172,3 +207,133 @@ def delete_image(image_uuid, client):
...
+            if subnet["status"]["name"] == subnet_name:
+                subnet_uuid.append(subnet["metadata"]["uuid"])
+
+        total_matches = subnet_list["metadata"]["total_matches"]
+        offset += length
+
+    return subnet_uuid
+
+
+def groups_call(filter, client):
```
Comment (AmitVasantSalunke) [inline plugins/module_utils/nutanix_api_client.py:374]: Please add doc string for function which will explain about purpose, in args, return value etc

File: plugins/module_utils/nutanix_api_client.py [hunk 5]
```diff
@@ -172,3 +207,133 @@ def delete_image(image_uuid, client):
...
+    sc_list_response = client.request(
+        api_endpoint="v3/groups", method="POST", data=json.dumps(filter))
+    return sc_list_response.json()
+
+
+def get_cluster_storage_container_map(storage_container_name, client):
```
Comment (AmitVasantSalunke) [inline plugins/module_utils/nutanix_api_client.py:388]: Please add doc string for function which will explain about purpose, in args, return value etc

File: plugins/module_utils/nutanix_api_client.py [hunk 6]
```diff
@@ -172,3 +207,133 @@ def delete_image(image_uuid, client):
...
+            if sc_name == storage_container_name:
+                cluster_sc_map[cluster] = entity_id
+
+        total_matches = sc_list["total_entity_count"]
+        offset += length
+
+    return cluster_sc_map
+
+
+def is_uuid(UUID):
```
Comment (AmitVasantSalunke) [inline plugins/module_utils/nutanix_api_client.py:432]: Please add doc string for function which will explain about purpose, in args, return value etc

File: plugins/module_utils/nutanix_api_client.py [hunk 7]
```diff
@@ -172,3 +207,133 @@ def delete_image(image_uuid, client):
...
+    try:
+        uuid.UUID(UUID)
+        return True
+    except ValueError:
+        return False
+
+
+def set_payload_keys(params, payload_format, payload):
```
Comment (AmitVasantSalunke) [inline plugins/module_utils/nutanix_api_client.py:447]: Please add doc string for function which will explain about purpose, in args, return value etc

File: plugins/module_utils/nutanix_api_client.py [hunk 8]
```diff
@@ -172,3 +207,133 @@ def delete_image(image_uuid, client):
...
+        elif type(params[i]) is list:
+            payload[i] = []
+            for item in params[i]:
+                payload[i].append(set_payload_keys(item, payload_format[i][0], {}))
+        elif type(params[i]) is str or type(params[i]) is int:
+            payload[i] = params[i]
+    return payload
+
+
+def has_changed(source_payload, destination_payload):
```
Comment (AmitVasantSalunke) [inline plugins/module_utils/nutanix_api_client.py:472]: Please add doc string for function which will explain about purpose, in args, return value etc

File: plugins/module_utils/nutanix_api_client.py [hunk 9]
```diff
@@ -172,3 +207,133 @@ def delete_image(image_uuid, client):
...
+        elif type(source_payload[key]) is str or type(source_payload[key]) is int:
+            if source_payload[key] != destination_payload[key]:
+                return True
+
+        if status:
+            return status
+    return status
+
+
+def read_file(filename):
```
Comment (AmitVasantSalunke) [inline plugins/module_utils/nutanix_api_client.py:500]: Please add doc string for function which will explain about purpose, in args, return value etc

File: plugins/modules/nutanix_vm.py [hunk 10]
```diff
@@ -385,69 +821,146 @@ def entry_point(module, client):
     return func(module.params, client)
 
 
-def create_vm_spec(params, vm_spec):
+def create_vm_spec(params, vm_spec, client):
```
Comment (AmitVasantSalunke) [inline plugins/modules/nutanix_vm.py:828]: Please add doc string for function which will explain about purpose, in args, return value etc

File: plugins/modules/nutanix_vm.py [hunk 11]
```diff
@@ -466,108 +985,91 @@ def create_vm_spec(params, vm_spec):
...
             }
 
-    vm_spec["spec"]["cluster_reference"] = {"kind": "cluster", "uuid": params['cluster']}
+    return vm_spec, None
 
-    return vm_spec
 
+def update_vm_spec(params, current_vm_payload, client):
 
-def update_vm_spec(params, vm_data):
```
Comment (AmitVasantSalunke) [inline plugins/modules/nutanix_vm.py:474]: Please add doc string for function which will explain about purpose, in args, return value etc

**Review by AmitVasantSalunke — COMMENTED** [review]
please update accordingly

File: plugins/module_utils/nutanix_api_client.py (hunk 1 above)
Comment (saratkumarkarri) [reply plugins/module_utils/nutanix_api_client.py]: done

File: plugins/module_utils/nutanix_api_client.py (hunk 6 above)
Comment (saratkumarkarri) [reply plugins/module_utils/nutanix_api_client.py:432]: done added doc string

**Review by balugeorge — APPROVED** [review]

**Review by AmitVasantSalunke — APPROVED** [review]
LGTM

**Pull request merged**